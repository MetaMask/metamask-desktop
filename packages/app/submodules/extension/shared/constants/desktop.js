export const EXTENSION_ERROR_PAGE_TYPES = {
  NOT_FOUND: 'not-found',
  CONNECTION_LOST: 'connection-lost',
  DESKTOP_OUTDATED: 'desktop-outdated',
  EXTENSION_OUTDATED: 'extension-outdated',
  CRITICAL_ERROR: 'critical-error',
};

export const DESKTOP_HOOK_TYPES = {
  DISCONNECT: 'disconnect',
};
