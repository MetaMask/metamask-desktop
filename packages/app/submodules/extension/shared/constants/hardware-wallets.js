/**
 * Accounts can be instantiated from simple, HD or the multiple hardware wallet
 * keyring types. Both simple and HD are treated as default but we do special
 * case accounts managed by a hardware wallet.
 */
export const KEYRING_TYPES = {
  LEDGER: 'Ledger Hardware',
  TREZOR: 'Trezor Hardware',
  LATTICE: 'Lattice Hardware',
  QR: 'QR Hardware Wallet Device',
  IMPORTED: 'Simple Key Pair',
};

export const DEVICE_NAMES = {
  LEDGER: 'ledger',
  TREZOR: 'trezor',
  QR: 'QR Hardware',
  LATTICE: 'lattice',
};

export const KEYRING_NAMES = {
  LEDGER: 'Ledger',
  TREZOR: 'Trezor',
  QR: 'QR',
  LATTICE: 'Lattice1',
};

/**
 * Used for setting the users preference for ledger transport type
 */
export const LEDGER_TRANSPORT_TYPES = {
  LIVE: 'ledgerLive',
  WEBHID: 'webhid',
  U2F: 'u2f',
};

export const LEDGER_USB_VENDOR_ID = '0x2c97';

export const WEBHID_CONNECTED_STATUSES = {
  CONNECTED: 'connected',
  NOT_CONNECTED: 'notConnected',
  UNKNOWN: 'unknown',
};

export const TRANSPORT_STATES = {
  NONE: 'NONE',
  VERIFIED: 'VERIFIED',
  DEVICE_OPEN_FAILURE: 'DEVICE_OPEN_FAILURE',
  UNKNOWN_FAILURE: 'UNKNOWN_FAILURE',
};

export const AFFILIATE_LINKS = {
  LEDGER: 'https://shop.ledger.com/?r=17c4991a03fa',
  GRIDPLUS: 'https://gridplus.io/?afmc=7p',
  TREZOR:
    'https://shop.trezor.io/product/trezor-one-black?offer_id=35&aff_id=11009',
  KEYSTONE:
    'https://shop.keyst.one/?rfsn=6088257.656b3e9&utm_source=refersion&utm_medium=affiliate&utm_campaign=6088257.656b3e9',
  AIRGAP: 'https://airgap.it/',
  COOLWALLET: 'https://www.coolwallet.io/',
  DCENT: 'https://dcentwallet.com/',
};

export const AFFILIATE_TUTORIAL_LINKS = {
  LEDGER:
    'https://support.ledger.com/hc/en-us/articles/4404366864657-Set-up-and-use-MetaMask-to-access-your-Ledger-Ethereum-ETH-account?docs=true',
  GRIDPLUS: 'https://docs.gridplus.io/setup/metamask',
  TREZOR: 'https://wiki.trezor.io/Apps:MetaMask',
  KEYSTONE:
    'https://support.keyst.one/3rd-party-wallets/eth-and-web3-wallets-keystone/bind-metamask-with-keystone',
  AIRGAP: 'https://support.airgap.it/guides/metamask/',
  COOLWALLET: 'https://www.coolwallet.io/metamask-step-by-step-guides/',
  DCENT:
    'https://medium.com/dcentwallet/dcent-wallet-now-supports-qr-based-protocol-to-link-with-metamask-57555f02603f',
};
