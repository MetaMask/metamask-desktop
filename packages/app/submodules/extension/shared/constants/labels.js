// The character limit on ENS names, nicknames and addresses before we truncate
export const TRUNCATED_NAME_CHAR_LIMIT = 11;

// The number of characters to slice from the beginning of an address for truncated format:
// `${TRUNCATED_ADDRESS_START_CHARS}...${TRUNCATED_ADDRESS_END_CHARS}`
export const TRUNCATED_ADDRESS_START_CHARS = 5;

// The number of characters to slice from the end of an address for truncated format:
// `${TRUNCATED_ADDRESS_START_CHARS}...${TRUNCATED_ADDRESS_END_CHARS}`
export const TRUNCATED_ADDRESS_END_CHARS = 4;
