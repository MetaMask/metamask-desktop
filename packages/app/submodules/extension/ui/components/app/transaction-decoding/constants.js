const TX_INSIGHTS_BASE_URI = 'https://tx-insights.metaswap.codefi.network';

export const FETCH_PROJECT_INFO_URI = `${TX_INSIGHTS_BASE_URI}/fetch-project`;
export const FETCH_SUPPORTED_NETWORKS_URI = `${TX_INSIGHTS_BASE_URI}/networks`;
