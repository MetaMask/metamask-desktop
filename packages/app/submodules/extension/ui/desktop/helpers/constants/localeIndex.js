const localeIndex = [
  { code: 'de', name: 'Deutsch' },
  { code: 'en', name: 'English' },
];

export default localeIndex;
