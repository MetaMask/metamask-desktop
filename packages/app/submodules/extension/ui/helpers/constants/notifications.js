///: BEGIN:ONLY_INCLUDE_IN(flask,desktopextension,desktopapp)

// The time after which a notification should be deleted.
export const NOTIFICATIONS_EXPIRATION_DELAY = 10000;

///: END:ONLY_INCLUDE_IN
