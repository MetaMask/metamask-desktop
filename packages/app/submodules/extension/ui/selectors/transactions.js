import { createSelector } from 'reselect';
import {
  PRIORITY_STATUS_HASH,
  PENDING_STATUS_HASH,
} from '../helpers/constants/transactions';
import txHelper from '../helpers/utils/tx-helper';
import {
  TRANSACTION_STATUSES,
  TRANSACTION_TYPES,
  SMART_TRANSACTION_STATUSES,
} from '../../shared/constants/transaction';
import { transactionMatchesNetwork } from '../../shared/modules/transaction.utils';
import { hexToDecimal } from '../../shared/lib/metamask-controller-utils';
import {
  getCurrentChainId,
  deprecatedGetCurrentNetworkId,
  getSelectedAddress,
} from './selectors';

const INVALID_INITIAL_TRANSACTION_TYPES = [
  TRANSACTION_TYPES.CANCEL,
  TRANSACTION_TYPES.RETRY,
];

export const incomingTxListSelector = (state) => {
  const { showIncomingTransactions } = state.metamask.featureFlags;
  if (!showIncomingTransactions) {
    return [];
  }

  const {
    network,
    provider: { chainId },
  } = state.metamask;
  const selectedAddress = getSelectedAddress(state);
  return Object.values(state.metamask.incomingTransactions).filter(
    (tx) =>
      tx.txParams.to === selectedAddress &&
      transactionMatchesNetwork(tx, chainId, network),
  );
};
export const unapprovedMsgsSelector = (state) => state.metamask.unapprovedMsgs;
export const currentNetworkTxListSelector = (state) =>
  state.metamask.currentNetworkTxList;
export const unapprovedPersonalMsgsSelector = (state) =>
  state.metamask.unapprovedPersonalMsgs;
export const unapprovedDecryptMsgsSelector = (state) =>
  state.metamask.unapprovedDecryptMsgs;
export const unapprovedEncryptionPublicKeyMsgsSelector = (state) =>
  state.metamask.unapprovedEncryptionPublicKeyMsgs;
export const unapprovedTypedMessagesSelector = (state) =>
  state.metamask.unapprovedTypedMessages;

export const smartTransactionsListSelector = (state) =>
  state.metamask.smartTransactionsState?.smartTransactions?.[
    getCurrentChainId(state)
  ]
    ?.filter((stx) => !stx.confirmed)
    .map((stx) => ({
      ...stx,
      transactionType: TRANSACTION_TYPES.SMART,
      status: stx.status?.startsWith('cancelled')
        ? SMART_TRANSACTION_STATUSES.CANCELLED
        : stx.status,
    }));

export const selectedAddressTxListSelector = createSelector(
  getSelectedAddress,
  currentNetworkTxListSelector,
  smartTransactionsListSelector,
  (selectedAddress, transactions = [], smTransactions = []) => {
    return transactions
      .filter(({ txParams }) => txParams.from === selectedAddress)
      .concat(smTransactions);
  },
);

export const unapprovedMessagesSelector = createSelector(
  unapprovedMsgsSelector,
  unapprovedPersonalMsgsSelector,
  unapprovedDecryptMsgsSelector,
  unapprovedEncryptionPublicKeyMsgsSelector,
  unapprovedTypedMessagesSelector,
  deprecatedGetCurrentNetworkId,
  getCurrentChainId,
  (
    unapprovedMsgs = {},
    unapprovedPersonalMsgs = {},
    unapprovedDecryptMsgs = {},
    unapprovedEncryptionPublicKeyMsgs = {},
    unapprovedTypedMessages = {},
    network,
    chainId,
  ) =>
    txHelper(
      {},
      unapprovedMsgs,
      unapprovedPersonalMsgs,
      unapprovedDecryptMsgs,
      unapprovedEncryptionPublicKeyMsgs,
      unapprovedTypedMessages,
      network,
      chainId,
    ) || [],
);

export const transactionSubSelector = createSelector(
  unapprovedMessagesSelector,
  incomingTxListSelector,
  (unapprovedMessages = [], incomingTxList = []) => {
    return unapprovedMessages.concat(incomingTxList);
  },
);

export const transactionsSelector = createSelector(
  transactionSubSelector,
  selectedAddressTxListSelector,
  (subSelectorTxList = [], selectedAddressTxList = []) => {
    const txsToRender = selectedAddressTxList.concat(subSelectorTxList);

    return txsToRender.sort((a, b) => b.time - a.time);
  },
);

/**
 * @name insertOrderedNonce
 * @private
 * @description Inserts (mutates) a nonce into an array of ordered nonces, sorted in ascending
 * order.
 * @param {string[]} nonces - Array of nonce strings in hex
 * @param {string} nonceToInsert - Nonce string in hex to be inserted into the array of nonces.
 */
const insertOrderedNonce = (nonces, nonceToInsert) => {
  let insertIndex = nonces.length;

  for (let i = 0; i < nonces.length; i++) {
    const nonce = nonces[i];

    if (Number(hexToDecimal(nonce)) > Number(hexToDecimal(nonceToInsert))) {
      insertIndex = i;
      break;
    }
  }

  nonces.splice(insertIndex, 0, nonceToInsert);
};

/**
 * @name insertTransactionByTime
 * @private
 * @description Inserts (mutates) a transaction object into an array of ordered transactions, sorted
 * in ascending order by time.
 * @param {object[]} transactions - Array of transaction objects.
 * @param {object} transaction - Transaction object to be inserted into the array of transactions.
 */
const insertTransactionByTime = (transactions, transaction) => {
  const { time } = transaction;

  let insertIndex = transactions.length;

  for (let i = 0; i < transactions.length; i++) {
    const tx = transactions[i];

    if (tx.time > time) {
      insertIndex = i;
      break;
    }
  }

  transactions.splice(insertIndex, 0, transaction);
};

/**
 * Contains transactions and properties associated with those transactions of the same nonce.
 *
 * @typedef {object} transactionGroup
 * @property {string} nonce - The nonce that the transactions within this transactionGroup share.
 * @property {object[]} transactions - An array of transaction (txMeta) objects.
 * @property {object} initialTransaction - The transaction (txMeta) with the lowest "time".
 * @property {object} primaryTransaction - Either the latest transaction or the confirmed
 * transaction.
 * @property {boolean} hasRetried - True if a transaction in the group was a retry transaction.
 * @property {boolean} hasCancelled - True if a transaction in the group was a cancel transaction.
 */

/**
 * @name insertTransactionGroupByTime
 * @private
 * @description Inserts (mutates) a transactionGroup object into an array of ordered
 * transactionGroups, sorted in ascending order by nonce.
 * @param {transactionGroup[]} transactionGroups - Array of transactionGroup objects.
 * @param {transactionGroup} transactionGroup - transactionGroup object to be inserted into the
 * array of transactionGroups.
 */
const insertTransactionGroupByTime = (transactionGroups, transactionGroup) => {
  const { primaryTransaction: { time: groupToInsertTime } = {} } =
    transactionGroup;

  let insertIndex = transactionGroups.length;

  for (let i = 0; i < transactionGroups.length; i++) {
    const txGroup = transactionGroups[i];
    const { primaryTransaction: { time } = {} } = txGroup;

    if (time > groupToInsertTime) {
      insertIndex = i;
      break;
    }
  }

  transactionGroups.splice(insertIndex, 0, transactionGroup);
};

/**
 * @name mergeNonNonceTransactionGroups
 * @private
 * @description Inserts (mutates) transactionGroups that are not to be ordered by nonce into an array
 * of nonce-ordered transactionGroups by time.
 * @param {transactionGroup[]} orderedTransactionGroups - Array of transactionGroups ordered by
 * nonce.
 * @param {transactionGroup[]} nonNonceTransactionGroups - Array of transactionGroups not intended to be ordered by nonce,
 * but intended to be ordered by timestamp
 */
const mergeNonNonceTransactionGroups = (
  orderedTransactionGroups,
  nonNonceTransactionGroups,
) => {
  nonNonceTransactionGroups.forEach((transactionGroup) => {
    insertTransactionGroupByTime(orderedTransactionGroups, transactionGroup);
  });
};

/**
 * @name nonceSortedTransactionsSelector
 * @description Returns an array of transactionGroups sorted by nonce in ascending order.
 * @returns {transactionGroup[]}
 */
export const nonceSortedTransactionsSelector = createSelector(
  transactionsSelector,
  (transactions = []) => {
    const unapprovedTransactionGroups = [];
    const incomingTransactionGroups = [];
    const orderedNonces = [];
    const nonceToTransactionsMap = {};

    transactions.forEach((transaction) => {
      const {
        txParams: { nonce } = {},
        status,
        type,
        time: txTime,
        txReceipt,
      } = transaction;

      if (typeof nonce === 'undefined' || type === TRANSACTION_TYPES.INCOMING) {
        const transactionGroup = {
          transactions: [transaction],
          initialTransaction: transaction,
          primaryTransaction: transaction,
          hasRetried: false,
          hasCancelled: false,
          nonce,
        };

        if (type === TRANSACTION_TYPES.INCOMING) {
          incomingTransactionGroups.push(transactionGroup);
        } else {
          insertTransactionGroupByTime(
            unapprovedTransactionGroups,
            transactionGroup,
          );
        }
      } else if (nonce in nonceToTransactionsMap) {
        const nonceProps = nonceToTransactionsMap[nonce];
        insertTransactionByTime(nonceProps.transactions, transaction);

        const {
          primaryTransaction: { time: primaryTxTime = 0 } = {},
          initialTransaction: { time: initialTxTime = 0 } = {},
        } = nonceProps;

        // Current Transaction Logic Cases
        // --------------------------------------------------------------------
        // Current transaction: The transaction we are examining in this loop.
        // Each iteration should be in time order, but that is not guaranteed.
        // --------------------------------------------------------------------
        const currentTransaction = {
          // A on chain failure means the current transaction was submitted and
          // considered for inclusion in a block but something prevented it
          // from being included, such as slippage on gas prices and conversion
          // when doing a swap. These transactions will have a '0x0' value in
          // the txReceipt.status field.
          isOnChainFailure: txReceipt?.status === '0x0',
          // Another type of failure is a "off chain" or "network" failure,
          // where the error occurs on the JSON RPC call to the network client
          // (Like Infura). These transactions are never broadcast for
          // inclusion and the nonce associated with them is not consumed. When
          // this occurs  the next transaction will have the same nonce as the
          // current, failed transaction. A failed on chain transaction will
          // not have the FAILED status although it should (future TODO: add a
          // new FAILED_ON_CHAIN) status. I use the word "Ephemeral" here
          // because a failed transaction that does not get broadcast is not
          // known outside of the user's local MetaMask and the nonce
          // associated will be applied to the next.
          isEphemeral:
            status === TRANSACTION_STATUSES.FAILED &&
            txReceipt?.status !== '0x0',
          // We never want to use a speed up (retry) or cancel as the initial
          // transaction in a group, regardless of time order. This is because
          // useTransactionDisplayData cannot parse a retry or cancel because
          // it lacks information on whether its a simple send, token transfer,
          // etc.
          isRetryOrCancel: INVALID_INITIAL_TRANSACTION_TYPES.includes(type),
          // Primary transactions usually are the latest transaction by time,
          // but not always. This value shows whether this transaction occurred
          // after the current primary.
          occurredAfterPrimary: txTime > primaryTxTime,
          // Priority Statuses are those that are ones either already confirmed
          // on chain, submitted to the network, or waiting for user approval.
          // These statuses typically indicate a transaction that needs to have
          // its status reflected in the UI.
          hasPriorityStatus: status in PRIORITY_STATUS_HASH,
          // A confirmed transaction is the most valid transaction status to
          // display because no other transaction of the same nonce can have a
          // more valid status.
          isConfirmed: status === TRANSACTION_STATUSES.CONFIRMED,
          // Initial transactions usually are the earliest transaction by time,
          // but not always. THis value shows whether this transaction occurred
          // before the current initial.
          occurredBeforeInitial: txTime < initialTxTime,
          // We only allow users to retry the transaction in certain scenarios
          // to help shield from expensive operations and other unwanted side
          // effects. This value is used to determine if the entire transaction
          // group should be marked as having had a retry.
          isValidRetry:
            type === TRANSACTION_TYPES.RETRY &&
            (status in PRIORITY_STATUS_HASH ||
              status === TRANSACTION_STATUSES.DROPPED),
          // We only allow users to cancel the transaction in certain scenarios
          // to help shield from expensive operations and other unwanted side
          // effects. This value is used to determine if the entire transaction
          // group should be marked as having had a cancel.
          isValidCancel:
            type === TRANSACTION_TYPES.CANCEL &&
            (status in PRIORITY_STATUS_HASH ||
              status === TRANSACTION_STATUSES.DROPPED),
        };

        // We should never assign a retry or cancel transaction as the initial,
        // likewise an ephemeral transaction should not be initial.
        currentTransaction.eligibleForInitial =
          !currentTransaction.isRetryOrCancel &&
          !currentTransaction.isEphemeral;

        // If a transaction failed on chain or was confirmed then it should
        // always be the primary because no other transaction is more valid.
        currentTransaction.shouldBePrimary =
          currentTransaction.isConfirmed || currentTransaction.isOnChainFailure;

        // Primary Transaction Logic Cases
        // --------------------------------------------------------------------
        // Primary transaction: The transaction for any given nonce which has
        // the most valid status on the network.
        // Example:
        // 1. Submit transaction A
        // 2. Speed up Transaction A.
        // 3. This creates a new Transaction (B) with higher gas params.
        // 4. Transaction A and Transaction B are both submitted.
        // 5. We expect Transaction B to be the most valid transaction to use
        //    for the status of the transaction group because it has higher
        //    gas params and should be included first.
        // The following logic variables are used for edge cases that protect
        // against UI bugs when this breaks down.
        const previousPrimaryTransaction = {
          // As we loop through the transactions in state we may temporarily
          // assign a primaryTransaction that is an "Ephemeral" transaction,
          // which is one that failed before being broadcast for inclusion in a
          // block. When this happens, and we have another transaction to
          // consider in a nonce group, we should use the new transaction.
          isEphemeral:
            nonceProps.primaryTransaction.status ===
              TRANSACTION_STATUSES.FAILED &&
            nonceProps.primaryTransaction?.txReceipt?.status !== '0x0',
        };

        // Initial Transaction Logic Cases
        // --------------------------------------------------------------------
        // Initial Transaction: The transaciton that most likely represents the
        // user's intent when creating/approving the transaction. In most cases
        // this is the first transaction of a nonce group, by time, but this
        // breaks down in the case of users with the advanced setting enabled
        // to set their own nonces manually. In that case a user may submit two
        // completely different transactions of the same nonce and they will be
        // bundled together by this selector as the same activity entry.
        const previousInitialTransaction = {
          // As we loop through the transactions in state we may temporarily
          // assign a initialTransaction that is an "Ephemeral" transaction,
          // which is one that failed before being broadcast for inclusion in a
          // block. When this happens, and we have another transaction to
          // consider in a nonce group, we should use the new transaction.
          isEphemeral:
            nonceProps.initialTransaction.status ===
              TRANSACTION_STATUSES.FAILED &&
            nonceProps.initialTransaction.txReceipt?.status !== '0x0',
        };

        // Check the above logic cases and assign a new primaryTransaction if
        // appropriate
        if (
          currentTransaction.shouldBePrimary ||
          previousPrimaryTransaction.isEphemeral ||
          (currentTransaction.occurredAfterPrimary &&
            currentTransaction.hasPriorityStatus)
        ) {
          nonceProps.primaryTransaction = transaction;
        }

        // Check the above logic cases and assign a new initialTransaction if
        // appropriate
        if (
          (currentTransaction.occurredBeforeInitial &&
            currentTransaction.eligibleForInitial) ||
          (previousInitialTransaction.isEphemeral &&
            currentTransaction.eligibleForInitial)
        ) {
          nonceProps.initialTransaction = transaction;
        }

        if (currentTransaction.isValidRetry) {
          nonceProps.hasRetried = true;
        }

        if (currentTransaction.isValidCancel) {
          nonceProps.hasCancelled = true;
        }
      } else {
        nonceToTransactionsMap[nonce] = {
          nonce,
          transactions: [transaction],
          initialTransaction: transaction,
          primaryTransaction: transaction,
          hasRetried:
            transaction.type === TRANSACTION_TYPES.RETRY &&
            (transaction.status in PRIORITY_STATUS_HASH ||
              transaction.status === TRANSACTION_STATUSES.DROPPED),
          hasCancelled:
            transaction.type === TRANSACTION_TYPES.CANCEL &&
            (transaction.status in PRIORITY_STATUS_HASH ||
              transaction.status === TRANSACTION_STATUSES.DROPPED),
        };

        insertOrderedNonce(orderedNonces, nonce);
      }
    });

    const orderedTransactionGroups = orderedNonces.map(
      (nonce) => nonceToTransactionsMap[nonce],
    );
    mergeNonNonceTransactionGroups(
      orderedTransactionGroups,
      incomingTransactionGroups,
    );
    return unapprovedTransactionGroups
      .concat(orderedTransactionGroups)
      .map((txGroup) => {
        // In the case that we have a cancel or retry as initial transaction
        // and there is a valid transaction in the group, we should reassign
        // the other valid transaction as initial. In this case validity of the
        // transaction is expanded to include off-chain failures because it is
        // valid to retry those with higher gas prices.
        if (
          INVALID_INITIAL_TRANSACTION_TYPES.includes(
            txGroup.initialTransaction?.type,
          )
        ) {
          const nonRetryOrCancel = txGroup.transactions.find(
            (tx) => !INVALID_INITIAL_TRANSACTION_TYPES.includes(tx.type),
          );
          if (nonRetryOrCancel) {
            return {
              ...txGroup,
              initialTransaction: nonRetryOrCancel,
            };
          }
        }
        return txGroup;
      });
  },
);

/**
 * @name nonceSortedPendingTransactionsSelector
 * @description Returns an array of transactionGroups where transactions are still pending sorted by
 * nonce in descending order.
 * @returns {transactionGroup[]}
 */
export const nonceSortedPendingTransactionsSelector = createSelector(
  nonceSortedTransactionsSelector,
  (transactions = []) =>
    transactions.filter(
      ({ primaryTransaction }) =>
        primaryTransaction.status in PENDING_STATUS_HASH,
    ),
);

/**
 * @name nonceSortedCompletedTransactionsSelector
 * @description Returns an array of transactionGroups where transactions are confirmed sorted by
 * nonce in descending order.
 * @returns {transactionGroup[]}
 */
export const nonceSortedCompletedTransactionsSelector = createSelector(
  nonceSortedTransactionsSelector,
  (transactions = []) =>
    transactions
      .filter(
        ({ primaryTransaction }) =>
          !(primaryTransaction.status in PENDING_STATUS_HASH),
      )
      .reverse(),
);

export const submittedPendingTransactionsSelector = createSelector(
  transactionsSelector,
  (transactions = []) =>
    transactions.filter(
      (transaction) => transaction.status === TRANSACTION_STATUSES.SUBMITTED,
    ),
);
