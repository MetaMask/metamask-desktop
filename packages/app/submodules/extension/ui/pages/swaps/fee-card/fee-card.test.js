import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { useSelector } from 'react-redux';

import {
  renderWithProvider,
  createSwapsMockStore,
  setBackgroundConnection,
  MOCKS,
  fireEvent,
} from '../../../../test/jest';
import { CHAIN_IDS } from '../../../../shared/constants/network';

import { checkNetworkAndAccountSupports1559 } from '../../../selectors';
import {
  getGasEstimateType,
  getGasFeeEstimates,
  getIsGasEstimatesLoading,
} from '../../../ducks/metamask/metamask';
import { TRANSACTION_ENVELOPE_TYPE_NAMES } from '../../../helpers/constants/transactions';

import FeeCard from '.';

const middleware = [thunk];

jest.mock('../../../hooks/useGasFeeEstimates', () => ({
  useGasFeeEstimates: jest.fn(),
}));

jest.mock('react-redux', () => {
  const actual = jest.requireActual('react-redux');

  return {
    ...actual,
    useSelector: jest.fn(),
  };
});

const generateUseSelectorRouter = () => (selector) => {
  if (selector === checkNetworkAndAccountSupports1559) {
    return true;
  }
  if (selector === getGasEstimateType) {
    return TRANSACTION_ENVELOPE_TYPE_NAMES.FEE_MARKET;
  }
  if (selector === getGasFeeEstimates) {
    return MOCKS.createGasFeeEstimatesForFeeMarket();
  }
  if (selector === getIsGasEstimatesLoading) {
    return false;
  }
  return undefined;
};

setBackgroundConnection({
  getGasFeeTimeEstimate: jest.fn(),
  getGasFeeEstimatesAndStartPolling: jest.fn(),
  createTransactionEventFragment: jest.fn(),
});

const createProps = (customProps = {}) => {
  return {
    primaryFee: {
      fee: '0.0441 ETH',
      maxFee: '0.04851 ETH',
    },
    secondaryFee: {
      fee: '$101.98',
      maxFee: '$112.17',
    },
    hideTokenApprovalRow: false,
    onFeeCardMaxRowClick: jest.fn(),
    tokenApprovalTextComponent: (
      <span key="fee-card-approve-symbol" className="fee-card__bold">
        ABC
      </span>
    ),
    tokenApprovalSourceTokenSymbol: 'ABC',
    onTokenApprovalClick: jest.fn(),
    metaMaskFee: '0.875',
    isBestQuote: true,
    numberOfQuotes: 6,
    onQuotesClick: jest.fn(),
    tokenConversionRate: 0.015,
    chainId: CHAIN_IDS.MAINNET,
    networkAndAccountSupports1559: false,
    supportsEIP1559V2: false,
    ...customProps,
  };
};

describe('FeeCard', () => {
  it('renders the component with initial props', () => {
    useSelector.mockImplementation(generateUseSelectorRouter());
    const props = createProps();
    const { getByText } = renderWithProvider(<FeeCard {...props} />);
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText('Max fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(getByText('Includes a 0.875% MetaMask fee.')).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
    expect(document.querySelector('.info-tooltip')).toMatchSnapshot();
    expect(getByText('Edit limit')).toBeInTheDocument();
  });

  it('renders the component with EIP-1559 enabled', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      networkAndAccountSupports1559: true,
      maxPriorityFeePerGasDecGWEI: '3',
      maxFeePerGasDecGWEI: '4',
    });
    const { getByText } = renderWithProvider(<FeeCard {...props} />, store);
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText('Max fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(getByText('Includes a 0.875% MetaMask fee.')).toBeInTheDocument();
    expect(
      document.querySelector('.fee-card__top-bordered-row'),
    ).toMatchSnapshot();
  });

  it('renders the component with smart transactions enabled and user opted in', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      smartTransactionsOptInStatus: true,
      smartTransactionsEnabled: true,
      maxPriorityFeePerGasDecGWEI: '3',
      maxFeePerGasDecGWEI: '4',
    });
    const { getByText, queryByTestId } = renderWithProvider(
      <FeeCard {...props} />,
      store,
    );
    expect(getByText('Best of 6 quotes.')).toBeInTheDocument();
    expect(getByText('Estimated gas fee')).toBeInTheDocument();
    expect(getByText(props.primaryFee.fee)).toBeInTheDocument();
    expect(getByText(props.secondaryFee.fee)).toBeInTheDocument();
    expect(getByText(`: ${props.secondaryFee.maxFee}`)).toBeInTheDocument();
    expect(queryByTestId('fee-card__edit-link')).not.toBeInTheDocument();
  });

  it('renders the component with hidden token approval row', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      hideTokenApprovalRow: true,
    });
    const { queryByText } = renderWithProvider(<FeeCard {...props} />, store);
    expect(queryByText('Edit limit')).not.toBeInTheDocument();
  });

  it('approves a token', () => {
    const store = configureMockStore(middleware)(createSwapsMockStore());
    const props = createProps({
      onTokenApprovalClick: jest.fn(),
    });
    const { queryByText } = renderWithProvider(<FeeCard {...props} />, store);
    fireEvent.click(queryByText('Edit limit'));
    expect(props.onTokenApprovalClick).toHaveBeenCalled();
  });
});
