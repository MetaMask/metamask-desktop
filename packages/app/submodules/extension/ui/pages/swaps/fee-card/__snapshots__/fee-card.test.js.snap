// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeCard renders the component with EIP-1559 enabled 1`] = `null`;

exports[`FeeCard renders the component with initial props 1`] = `null`;

exports[`FeeCard renders the component with initial props 2`] = `
<div
  class="info-tooltip"
>
  <div
    class="fee-card__row-label fee-card__info-tooltip-container"
  >
    <div
      aria-describedby="tippy-tooltip-1"
      class="info-tooltip__tooltip-container fee-card__info-tooltip-content-container"
      data-original-title="null"
      data-tooltipped=""
      style="display: inline;"
      tabindex="0"
    >
      <svg
        viewBox="0 0 10 10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 0C2.2 0 0 2.2 0 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 2c.4 0 .7.3.7.7s-.3.7-.7.7-.7-.2-.7-.6.3-.8.7-.8zm.7 6H4.3V4.3h1.5V8z"
          fill="var(--color-icon-alternative)"
        />
      </svg>
    </div>
  </div>
</div>
`;
